### PR TITLE
[stdlib] Swift4 Modernizations Compatible with Swift 3.2

### PR DESCRIPTION
**Pull request by dabrahams**
Applies most of the changes to make the standard library build under Swift 4 mode, but doesn't throw the switch.  See #10981, which isn't ready for merge, for the whole megillah.  Whether this is ready is arguable; it could use review even if all the tests pass.

### CODE REVIEW

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by dabrahams** [comment]
@swift-ci Please test source compatibility

**Comment by dabrahams** [comment]
I believe #10994 will make these tests pass

**Comment by dabrahams** [comment]
@swift-ci Please test source compatibility

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by dabrahams** [comment]
@moiseev @airspeedswift Can I get a review?  We really should apply these changes ASAP.  The source compatibility failure is an unrelated UPASS.

File: stdlib/public/core/StringRangeReplaceableCollection.swift.gyb [hunk 1]
```diff
@@ -44,7 +44,7 @@ extension String : StringProtocol, RangeReplaceableCollection {
   
   // This initializer satisfies the LosslessStringConvertible conformance
   @available(swift, obsoleted: 4, message: "String.init(_:String) is no longer failable")
-  public init?(_ other: String) {
+  public init?(_ other: String, obsoletedInSwift4: () = ()) {
```
Comment (milseman) [inline stdlib/public/core/StringRangeReplaceableCollection.swift.gyb]: Could you explain what this change is for?

**Review by milseman — COMMENTED** [review]
It all seems reasonable to me, but I do have a question as to why `String.init?(_)` became `String.init?(_:obsoletedInSwift4:)`.

File: stdlib/public/core/StringRangeReplaceableCollection.swift.gyb (hunk 1 above)
Comment (airspeedswift) [inline stdlib/public/core/StringRangeReplaceableCollection.swift.gyb]: Was going to ask same thing, other than this patch lgtm.

Comment (dabrahams) [inline stdlib/public/core/StringRangeReplaceableCollection.swift.gyb]: This hack prevents the `init?` from being used to satisfy a protocol requirement; you can satisfy a protocol requirement with something that's not yet available, but not with something that has become unavailable.  IOW, without this, we can't throw the switch on the Swift 4 transition.  I should add a comment to make that clear.

**Comment by dabrahams** [comment]
@swift-ci Please smoke test

**Comment by dabrahams** [comment]
@swift-ci Please smoke test and merge

**Comment by dabrahams** [comment]
@swift-ci Please smoke test and merge

**Pull request merged**